main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -1299,6 +1296,14 @@ object Typer {
           resultEff = declaredEff.getOrElse(actualEff)
         } yield (constrs, resultTyp, resultPur, resultEff)
 
+      case KindedAst.Expression.Upcast(exp, _, _, _, loc) =>
+        for {
```

Comment:
I think this should be:

```
       for {
          (constrs, _tpe, _pur, _eff) <- visitExp(exp)
        } yield {
          (constrs, tvar, pvar, evar)
        }
```

where tvar, pvar, and evar come from the upcast.

Reply:
Then in safety you have two things. The Upcast expression now knows the "expected" type, purity and effect of the "outer" whereas the "inner" expression has the "actual" type, purity, and effect.

Reply:
Alright, with this change it works!

Reply:
Hopefully, it also works when the safety error is fully implemented hehe.

Reply:
OK, and then in safety you can extract "actualTpe, actualPur, actualEff" and "expectedTpe, ..." and compare the,